### PR TITLE
write markdown tables in RST file

### PR DESCRIPTION
**Pull request by nathanesau**
tldr; We can now use markdown table syntax in RST file. Command for building docs is same as before (i.e. ``cmake -G \"Ninja Multi-Config\" -B build -DDOCS:STRING=YES && ninja -C build -f build-Release.ninja``). This was achieved by modifying cmake docs command from `` ${SPHINX_EXECUTABLE} -b html`` to ``${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/scripts/preprocessor.py --input ${SPHINX_SOURCE} --output ${SPHINX_SOURCE}/__output__ && ${SPHINX_EXECUTABLE} -b html``. Preprocessor may be useful for other custom RST syntax in future. Additional RST parsing can be added to ``docs/scripts/preprocessor.py`` in future.

Resolves https://github.com/codereport/jsource/issues/120.

Now, we can write tables in RST file using markdown syntax. This is pretty much as simple as it can get re. writing the tables. Supporting this requires additional step before the docs build, because markdown syntax isn't allowed in final RST file.

```rst

Verbs
=====
.. toctree::
.. contents::
   
*******
Monadic
*******

Tally
#####

.. doxygenfunction:: tally

******
Dyadic
******

******
Tables
******

Write markdown table in RST file.

| fruit | color | startswith |
| ----- | ----- | ---------- |
| apples | red | a |
| oranges | orange | o |
```

So, when building the docs, I added a preprocessor step (to existing CMAKE build docs command). The preprocessor scans all the RST files and creates parsed copy of each RST file in ``docs/source__out__``. I have added that folder to ``.gitignore``.

For above example, the parsed RST would be:

```rst

Verbs
=====
.. toctree::
.. contents::
   
*******
Monadic
*******

Tally
#####

.. doxygenfunction:: tally

******
Dyadic
******

******
Tables
******

Write markdown table in RST file.

======= ====== ==========
fruit   color  startswith
======= ====== ==========
apples  red    a         
oranges orange o         
======= ====== ==========
```

Notice that markdown table has been replaced with valid RST (type 2) table. 

Then after building you will get a nice table.

Screenshot_2021-02-07_13-04-13

### CODE REVIEW

File: docs/scripts/preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    row
+        | apples | red | a |
+
+    data
+        ['apples', 'red', 'a']
+    """
+    data = row.split("|")
+    data = data[0:len(data)-1]
+    data = [e.strip() for e in data if e.strip()]
+    return data
```
Comment (codereport) [inline docs/scripts/preprocessor.py:52]:
```suggestion
    return [e.strip() for e in row.split("|")][1:-1]
```

Comment (codereport) [reply docs/scripts/preprocessor.py:52]: This is an example of [ITM](https://www.youtube.com/watch?v=CjHgL5EQdcY). Are you able to try and refactor other places in the code do be more declarative?

**Review by codereport — CHANGES_REQUESTED** [review]

File: docs/scripts/preprocessor.py (hunk 1 above)
Comment (nathanesau) [reply docs/scripts/preprocessor.py:52]: Sure will do tomorrow. Thanks for the suggestion.

Comment (nathanesau) [reply docs/scripts/preprocessor.py]: I have refactored the code. I am pretty certain there are no more uses of ITM anymore. Hope this falls under the category of "beautiful" python refactoring :)

https://github.com/codereport/jsource/pull/129/commits/9b5356b596555b8730407fee8c94f5af29c8a27c

File: docs/scripts/preprocessor.py [hunk 2]
```diff
@@ -46,10 +41,7 @@ def get_markdown_row_data(row):
     data
         ['apples', 'red', 'a']
     """
-    data = row.split("|")
-    data = data[0:len(data)-1]
-    data = [e.strip() for e in data if e.strip()]
-    return data
+    return [e.strip() for e in row.split("|")[1:-1] if e.strip()]
```
Comment (codereport) [inline docs/scripts/preprocessor.py:44]: I don't think this works for blank cells, does it? i.e. `|cat|dog||`

File: docs/scripts/preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--input", help="Input Sphinx source directory")
+    parser.add_argument("--output", help="Output Sphinx source directory")
+    args = parser.parse_args()
+
+    try:
+        shutil.rmtree(args.output)
```
Comment (juntuu) [inline docs/scripts/preprocessor.py:129]: [`shutil.rmtree`](https://docs.python.org/3/library/shutil.html#shutil.rmtree) has optional argument `ignore_errors`, which could be quite nice for this use case

File: docs/scripts/preprocessor.py (hunk 2 above)
Comment (nathanesau) [reply docs/scripts/preprocessor.py]: fix checked in. you are right.

File: docs/scripts/preprocessor.py (hunk 3 above)
Comment (nathanesau) [reply docs/scripts/preprocessor.py]: i have made the change. thanks for the suggestion! always good to learn new python tricks.

**Review by codereport — APPROVED** [review]
LGTM, might refactor a couple things on next livestream
